cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -754,6 +839,30 @@ void RegisterScalarAggregateBasic(FunctionRegistry* registry) {
                aggregate::CountInit, func.get());
   DCHECK_OK(registry->AddFunction(std::move(func)));
 
+  func = std::make_shared<ScalarAggregateFunction>(
+      "count_distinct", Arity::Unary(), &count_distinct_doc, &default_count_options);
+
+  // Takes any input, outputs int64 scalar
+  aggregate::AddCountDistinctKernel<Int8Type>(int8(), func.get());
+  aggregate::AddCountDistinctKernel<Int16Type>(int16(), func.get());
+  aggregate::AddCountDistinctKernel<Int32Type>(int32(), func.get());
+  aggregate::AddCountDistinctKernel<Date32Type>(date32(), func.get());
+  aggregate::AddCountDistinctKernel<Int64Type>(int64(), func.get());
+  aggregate::AddCountDistinctKernel<UInt8Type>(uint8(), func.get());
+  aggregate::AddCountDistinctKernel<UInt16Type>(uint16(), func.get());
+  aggregate::AddCountDistinctKernel<UInt32Type>(uint32(), func.get());
+  aggregate::AddCountDistinctKernel<UInt64Type>(uint64(), func.get());
+  aggregate::AddCountDistinctKernel<FloatType>(float32(), func.get());
+  aggregate::AddCountDistinctKernel<DoubleType>(float64(), func.get());
+  aggregate::AddCountDistinctKernel<Time32Type>(time32(TimeUnit::SECOND), func.get());
+  aggregate::AddCountDistinctKernel<Time32Type>(time32(TimeUnit::MILLI), func.get());
+  aggregate::AddCountDistinctKernel<Time64Type>(time64(TimeUnit::MICRO), func.get());
+  aggregate::AddCountDistinctKernel<Time64Type>(time64(TimeUnit::NANO), func.get());
```

Comment:
You should take a look at how this is done for other kernels instead of enumerating types explicitly. For example, see `SumLikeInit`.

Reply:
Thanks I will check that example and see if I can improve the code here.